fix(goal_planner): correct the lane_id and velocity interpolation logic in smooth_goal_connection

yhisaki opened the pull request:
## Description

Before this PR, lane_id and velocity were computed using zero-order hold interpolation. However, this method fails to calculate correct values near lane boundaries. In this PR, the lane ID is calculated from map(route_handler), and the speed is calculated from the closest point on input_path.

**Before**

<img width="1466" height="899" alt="newplot (3)" src="https://github.com/user-attachments/assets/68dd1421-1efa-43cc-855c-c55c831b5aa3" />

**After**

<img width="1330" height="767" alt="newplot (2)" src="https://github.com/user-attachments/assets/78b09218-2434-4da5-8681-7999ca0bab2a" />

## Related links

**Parent Issue:**

- Link

<!-- ⬇️🟢
**Private Links:**

- [CompanyName internal link]()
⬆️🟢 -->

## How was this PR tested?

- [[PR check][Lexus][ControlDLRCommon_Basic] DLR test](https://evaluation.tier4.jp/evaluation/reports/c39c7f27-c07a-598c-92c3-3e2eb4b3d9bb?project_id=prd_jt)
- [[PR check][Lexus][FuncVerificationScenario_Basic] planning test](https://evaluation.tier4.jp/evaluation/reports/c196b389-61c4-549d-8885-be348d221eea?project_id=prd_jt)
- [[PR check][Lexus][CommonScenario_Basic] planning test](https://evaluation.tier4.jp/evaluation/reports/54ba3b0a-71fd-59ac-828a-3d104ab5cf9a?project_id=prd_jt)

## Notes for reviewers

None.

## Interface changes

None.

<!-- ⬇️🔴

### Topic changes

#### Additions and removals

| Change type   | Topic Type      | Topic Name    | Message Type        | Description       |
|:--------------|:----------------|:--------------|:--------------------|:------------------|
| Added/Removed | Pub/Sub/Srv/Cli | `/topic_name` | `std_msgs/String`   | Topic description |

#### Modifications

| Version | Topic Type      | Topic Name        | Message Type        | Description       |
|:--------|:----------------|:------------------|:--------------------|:------------------|
| Old     | Pub/Sub/Srv/Cli | `/old_topic_name` | `sensor_msgs/Image` | Topic description |
| New     | Pub/Sub/Srv/Cli | `/new_topic_name` | `sensor_msgs/Image` | Topic description |

### ROS Parameter Changes

#### Additions and removals

| Change type   | Parameter Name | Type     | Default Value | Description       |
|:--------------|:---------------|:---------|:--------------|:------------------|
| Added/Removed | `param_name`   | `double` | `1.0`         | Param description |

#### Modifications

| Version | Parameter Name   | Type     | Default Value | Description       |
|:--------|:-----------------|:---------|:--------------|:------------------|
| Old     | `old_param_name` | `double` | `1.0`         | Param description |
| New     | `new_param_name` | `double` | `1.0`         | Param description |

🔴⬆️ -->

## Effects on system behavior

None.

kosuke55 commented on planning/behavior_path_planner/autoware_behavior_path_planner_common/src/utils/utils.cpp:
```diff
@@ -250,11 +238,47 @@ std::optional<size_t> findIndexOutOfGoalSearchRange(
+  for (auto it = begin; it != end; ++it) {
+    const auto point = it->point;
+    lanelet::ConstLanelet lanelet;
+    if (route_handler->getClosestLaneletWithinRoute(point.pose, &lanelet)) {
```
nits:
this is edge case but may get wrong lane from overrlapped route lanes.
better to search from lane sequence after `min_dist_out_of_circle_index` lane id for continuity.

yhisaki committed:
feat(goal_planner): enhance goal refinement with route handler integration

- Updated `set_goal` and `refinePathForGoal` functions to include a `route_handler` parameter for improved path refinement.
- Introduced utility functions to fill lane IDs and longitudinal velocities based on the input path.
- Removed redundant lane ID filling logic to streamline goal setting process.

This change aims to enhance the accuracy of goal positioning and path planning by leveraging route information.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor(utils): reorder parameters in set_goal function for clarity

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor(goal_planner): enhance lanelet retrieval in path planning

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
fix(utils): enhance longitudinal velocity handling in fillLongitudinalVelocityFromInputPath

- Added a check to ensure the input path has at least two points before processing.
- Removed the last point from the input path to avoid zero velocity issues.
- Updated lane ID and longitudinal velocity assignment logic in set_goal to streamline the process.

These changes improve the robustness of the path planning by ensuring valid input and clearer handling of velocity data.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>

yhisaki committed:
refactor(utils): enhance lanelet handling in path processing

- Introduced a template function to retrieve unique lanelets from a given path, improving modularity and reusability.
- Updated `fillLaneIdsFromMap` to utilize the new lanelet retrieval function, streamlining lane ID assignment.
- Enhanced `set_goal` to incorporate the unique lanelets for more accurate lane ID filling.

These changes aim to improve the efficiency and clarity of lanelet management within the path planning process.

Signed-off-by: Y.Hisaki <yhisaki31@gmail.com>
